File: majority_element/majel.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

// more than ⌊ n/2 ⌋ times
int majorityElement(std::vector<int>& nums) {
  int n = nums.size();
  int half = n / 2;
  // std::cout << "half:" << half << '\n';
  std::unordered_map<int, int> numOccurrences;
  for(int i = 0; i < n; i++){
    int num = nums[i];
    if (numOccurrences.find(num) != numOccurrences.end()){
      numOccurrences[num] += 1;
    } else {
      numOccurrences[num] = 1;
    }
    if (numOccurrences[num] > half) return num;
    // std::cout << "numOccurrences[" << num << "]:" << numOccurrences[num] << '\n';
  }

  return nums[0]; // for 2 elements arrays
}

// c++ -std=c++11 majel.cpp && ./a.out
int main(int argc, char const *argv[]){
  // std::vector<int> nums{3,2,3}; // 3
  std::vector<int> nums{2,2,1,1,1,2,2}; // 2
  // std::vector<int> nums{2}; // 2
  // std::vector<int> nums{3, 2}; // 3

  int res = majorityElement(nums);

  std::cout << res << std::endl;

  return 0;
}

File: majority_element/Majel.java
import java.util.HashMap;
import java.util.Map;
import java.util.Arrays;

class Majel {
  public static int majorityElement(int[] nums) {
    // int half = nums.length / 2;
    // Map<Integer, Integer> numOccurrences = new HashMap<Integer, Integer>();

    // for (int num: nums){
    //   numOccurrences.compute(num, (k, v) -> (v == null) ? 1 : v + 1);
    //   if (numOccurrences.get(num) > half) return num;
    // }

    // Boyer-Moore
    int votes = 0;
    Integer candidate = null;
    for(int num : nums){
      if (votes == 0) candidate = num;
      votes += (num == candidate) ? +1 : -1;
    }
    return candidate;
  }

  public static void main(String[] args){
    // int[] nums = {3,2,3}; // 3
    // int[] nums = {3,2}; // 3
    // int[] nums = {3}; // 3
    int[] nums = {-1,-1,2147483647}; // -1

    int res = majorityElement(nums);

    String output = String.format("%s => %d", Arrays.toString(nums), res);
    System.out.println(output);
  }
}
